internal/k8s/reconciler/gateway.go

```diff
@@ -46,304 +43,85 @@ type K8sGatewayConfig struct {
-
-	for _, ingress := range updated.Status.LoadBalancer.Ingress {
-		g.GatewayState.ServiceReady = true
-		g.GatewayState.Addresses = append(g.GatewayState.Addresses, ingress.IP)
-	}
-
-	return nil
-}
-
-// assignGatewayIPFromService retrieves the internal cluster IP for the
-// Service and assigns it to the Gateway
-func (g *K8sGateway) assignGatewayIPFromService(ctx context.Context, service *corev1.Service) error {
-	updated, err := g.client.GetService(ctx, types.NamespacedName{Namespace: service.Namespace, Name: service.Name})
-	if err != nil {
-		return err
-	}
-
-	if updated == nil {
-		g.GatewayState.Status.Scheduled.NotReconciled = errors.New("service not found")
-		return nil
-	}
-
-	if updated.Spec.ClusterIP != "" {
-		g.GatewayState.ServiceReady = true
-		g.GatewayState.Addresses = append(g.GatewayState.Addresses, updated.Spec.ClusterIP)
+func (g *K8sGateway) ID() core.GatewayID {
+	return core.GatewayID{
+		Service:         g.Gateway.Name,
+		ConsulNamespace: g.consulNamespace,
 	}
-
-	return nil
 }
 
-// assignGatewayIPFromPod retrieves the internal IP for the Pod and assigns
-// it to the Gateway.
-func (g *K8sGateway) assignGatewayIPFromPod(ctx context.Context) error {
-	pod, err := g.client.PodWithLabels(ctx, utils.LabelsForGateway(g.Gateway))
-	if err != nil {
-		return err
-	}
-
-	if pod == nil {
-		g.GatewayState.Status.Scheduled.NotReconciled = errors.New("pod not found")
-		return nil
-	}
-
-	if pod.Status.PodIP != "" {
-		g.GatewayState.ServiceReady = true
-		g.GatewayState.Addresses = append(g.GatewayState.Addresses, pod.Status.PodIP)
+func (g *K8sGateway) Meta() map[string]string {
+	return map[string]string{
+		"external-source":                          "consul-api-gateway",
+		"consul-api-gateway/k8s/Gateway.Name":      g.Gateway.Name,
+		"consul-api-gateway/k8s/Gateway.Namespace": g.Gateway.Namespace,
 	}
-
-	return nil
 }
 
-// assignGatewayIPFromPodHost retrieves the (potentially) externally accessible
-// IP address for the host that the Pod is running on and assigns it to the Gateway.
-// This IP address is not always externally accessible and may require additional
-// work by the practitioner such as port-forwarding or opening firewall rules to make
-// it externally accessible.
-func (g *K8sGateway) assignGatewayIPFromPodHost(ctx context.Context) error {
-	pod, err := g.client.PodWithLabels(ctx, utils.LabelsForGateway(g.Gateway))
-	if err != nil {
-		return err
-	}
-
-	if pod == nil {
-		g.GatewayState.Status.Scheduled.NotReconciled = errors.New("pod not found")
+// Bind returns the name of the listeners to which a route bound
+func (g *K8sGateway) Bind(ctx context.Context, route store.Route) []string {
+	k8sRoute, ok := route.(*K8sRoute)
+	if !ok {
 		return nil
 	}
 
-	if pod.Status.HostIP != "" {
-		g.GatewayState.ServiceReady = true
-		g.GatewayState.Addresses = append(g.GatewayState.Addresses, pod.Status.HostIP)
-	}
-
-	return nil
+	return NewBinder(g.client, g.Gateway, g.GatewayState).Bind(ctx, k8sRoute)
```

Comment:
The Binder itself gets fully reworked in the 3rd PR to be re-usable across gateways/routes, but this is the interim state for now -- for some context, the movement towards extracting out all of the more primitive structures as arguments like the `*gw.Gateway` and `*state.GatewayState` was to be explicit about what was needed in what part of the binding/validation/resolution process just so I could more easily drop the fields once different parts of the refactor were complete.